src/drivers/driver_zephyr.c

```diff
@@ -1024,22 +1028,94 @@ static void wpa_drv_zep_event_signal_change(struct zep_drv_if_ctx *if_ctx,
 }
 
 static void wpa_drv_zep_event_roc_complete(struct zep_drv_if_ctx *if_ctx,
-					    int freq, unsigned int duration)
+					    int freq, unsigned int duration, u64 rpu_cookie)
 {
+	unsigned long long host_cookie = 0;
+	struct cookie_info *cookie_info = NULL;
+	struct cookie_info *tmp = NULL;
+	int match_cookie_found = 0;
 	union wpa_event_data event;
-	os_memset(&event, 0, sizeof(event));
-	event.remain_on_channel.freq = freq;
-	event.remain_on_channel.duration = duration;
-	wpa_supplicant_event_wrapper(if_ctx->supp_if_ctx, EVENT_REMAIN_ON_CHANNEL, &event);
+
+	if (dl_list_empty(&if_ctx->cookie_list)){
+		return;
+	}
+
+	dl_list_for_each_safe(cookie_info, tmp, &if_ctx->cookie_list, struct cookie_info, list) {
+		if (cookie_info) {
+			if (cookie_info->rpu_cookie != rpu_cookie) {
+				continue;
+			}
+			host_cookie = cookie_info->host_cookie;
+			match_cookie_found = 1;
+            break;
+        }
+	}
+
+	if (rpu_cookie != remain_on_chan_cookie) {
+		wpa_printf(MSG_DEBUG, "wpa_supp: No matching cookie found for ROC complete event");
+		return;
+	}
+
+	if (match_cookie_found) {
+		os_memset(&event, 0, sizeof(event));
+		event.remain_on_channel.freq = freq;
+		event.remain_on_channel.duration = duration;
+		wpa_supplicant_event_wrapper(if_ctx->supp_if_ctx, EVENT_REMAIN_ON_CHANNEL, &event);
+	}
 }
 
 static void wpa_drv_zep_event_roc_cancel_complete(struct zep_drv_if_ctx *if_ctx,
-						   int freq)
+						   int freq, u64 rpu_cookie)
 {
+	struct cookie_info *cookie_info = NULL;
+	struct cookie_info *tmp = NULL;
+	int match_cookie_found = 0;
 	union wpa_event_data event;
-	os_memset(&event, 0, sizeof(event));
-	event.remain_on_channel.freq = freq;
-	wpa_supplicant_event_wrapper(if_ctx->supp_if_ctx, EVENT_CANCEL_REMAIN_ON_CHANNEL, &event);
+
+	if (dl_list_empty(&if_ctx->cookie_list)) {
+		return;
+	}
+
+	dl_list_for_each_safe(cookie_info, tmp, &if_ctx->cookie_list, struct cookie_info, list) {
+		if (cookie_info) {
+			if (cookie_info->rpu_cookie != rpu_cookie) {
+				continue;
+			}
+			match_cookie_found = 1;
+			dl_list_del(&cookie_info->list);
+			os_free(cookie_info);
+            break;
+		}
+	}
+
+	if (match_cookie_found) {
+		os_memset(&event, 0, sizeof(event));
+		event.remain_on_channel.freq = freq;
+		wpa_supplicant_event_wrapper(if_ctx->supp_if_ctx,
+			EVENT_CANCEL_REMAIN_ON_CHANNEL, &event);
+	}
+}
+
+static void wpa_drv_zep_event_cookie_event(struct zep_drv_if_ctx *if_ctx,
+					  u64 host_cookie, u64 cookie)
+{
+	struct cookie_info *ck_info = NULL;
+
+	if (waiting_for_remain_on_channel_cookie) {
+		ck_info = os_zalloc(sizeof(*ck_info));
+		if (!ck_info) {
+			wpa_printf(MSG_ERROR, "wpa_supp: Failed to alloc cookie_info");
+			return;
+		}
+		ck_info->host_cookie = host_cookie;
+		ck_info->rpu_cookie = cookie;
+		dl_list_add(&if_ctx->cookie_list, &ck_info->list);
```

Comment:
The cookie are mainly to match responses to requests, and a single interface (can be physical or virtual depending on the underlying HW capability) can only have one RoC at a time, so, you don't need a global linked list here. Move the globals to a per-VIF structure and drop if not matched. See NL80211 driver for reference.